sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -604,16 +729,22 @@ sycl::detail::pi::PiProgram ProgramManager::getBuiltPIProgram(
     ContextImpl->addDeviceGlobalInitializer(BuiltProgram.get(), {Device}, &Img);
 
     // Save program to persistent cache if it is not there
-    if (!DeviceCodeWasInCache)
+    if (!DeviceCodeWasInCache) {
       PersistentDeviceCodeCache::putItemToDisc(
           Device, Img, SpecConsts, CompileOpts + LinkOpts, BuiltProgram.get());
+      // If we linked some extra device images, cache the same program for them
+      for (RTDeviceBinaryImage *BinImg : DeviceImagesToLink) {
+        PersistentDeviceCodeCache::putItemToDisc(Device, *BinImg, SpecConsts,
+                                                 CompileOpts + LinkOpts,
+                                                 BuiltProgram.get());
```

Comment:
Thinking about this more, we may not even need that. The situation with virtual functions is that there could be several kernels that use virtual function, but as soon as at least one of them is launched, they all will be linked together and only linked program will be used.

I think that we can safely assume that the order of kernels will most likely stay the same between program invocations and therefore it will always be the same device image used in the first cache lookup and for the rest of kernels we will re-use that same program. However, what I need to make sure that when we load cached program from disk, we should also ensure that it is properly stored in in-memory cache so it is re-used when second kernel is launched.

I will add corresponding TODOs and work on improving the situation in a follow-up patches.

Reply:
>  However, what I need to make sure that when we load cached program from disk, we should also ensure that it is properly stored in in-memory cache so it is re-used when second kernel is launched.

We do in-memory caching regardless of whether or not the program originated from the on-disk cache. Therefore, I will simply remove that extra persistent caching code.

Reply:
Removed this code in 628909e63bfab765253bf013fd1898df45df3a3b

Reply:
I'm not sure exactly how often that assumption holds true, but I think it's often enough that it's reasonable for the initial implementation.